From 5db39e2609f2bcc4c0a3ccca304f286e8ba6bcda Mon Sep 17 00:00:00 2001
From: Dentrax <furkan.turkal@chainguard.dev>
Date: Sun, 31 Aug 2025 17:43:03 +0300
Subject: [PATCH] entitlements: Grant entitlements for bc-fips 2.1 compat

Runtime.addShutdownHook is always forbidden, but new versions of
bc-fips and bc-rng-jent require it. There is no policy entitlement to
allow it, thus lift the forbidden policy.

bc-fips 2.1.0 has native libraries capability, thus grant loading
native libraries.

bc.rng.jent also uses native libraries, thus grant loading native
libraries to that scope.

Unconditionally grant these entitlements to allow running with bc-fips
in unapproved mode.

Note that the entitlements code is broken, because it does not correctly query security property of bouncycastle.

Signed-off-by: Dentrax <furkan.turkal@chainguard.dev>
Co-authored-by: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
---
 .../entitlement/bridge/EntitlementChecker.java    |  2 --
 .../entitlement/qa/test/SystemActions.java        |  5 -----
 .../initialization/HardcodedEntitlements.java     | 15 +++++++++++----
 .../api/ElasticsearchEntitlementChecker.java      |  5 -----
 4 files changed, 11 insertions(+), 16 deletions(-)

diff --git a/libs/entitlement/bridge/src/main/java/org/elasticsearch/entitlement/bridge/EntitlementChecker.java b/libs/entitlement/bridge/src/main/java/org/elasticsearch/entitlement/bridge/EntitlementChecker.java
index 74b5b987..22f4c7a3 100644
--- a/libs/entitlement/bridge/src/main/java/org/elasticsearch/entitlement/bridge/EntitlementChecker.java
+++ b/libs/entitlement/bridge/src/main/java/org/elasticsearch/entitlement/bridge/EntitlementChecker.java
@@ -175,8 +175,6 @@ public interface EntitlementChecker {
 
     void check$java_lang_System$$setOut(Class<?> callerClass, PrintStream out);
 
-    void check$java_lang_Runtime$addShutdownHook(Class<?> callerClass, Runtime runtime, Thread hook);
-
     void check$java_lang_Runtime$removeShutdownHook(Class<?> callerClass, Runtime runtime, Thread hook);
 
     void check$java_lang_Thread$$setDefaultUncaughtExceptionHandler(Class<?> callerClass, Thread.UncaughtExceptionHandler ueh);
diff --git a/libs/entitlement/qa/entitlement-test-plugin/src/main/java/org/elasticsearch/entitlement/qa/test/SystemActions.java b/libs/entitlement/qa/entitlement-test-plugin/src/main/java/org/elasticsearch/entitlement/qa/test/SystemActions.java
index c1281f50..b069e722 100644
--- a/libs/entitlement/qa/entitlement-test-plugin/src/main/java/org/elasticsearch/entitlement/qa/test/SystemActions.java
+++ b/libs/entitlement/qa/entitlement-test-plugin/src/main/java/org/elasticsearch/entitlement/qa/test/SystemActions.java
@@ -54,11 +54,6 @@ class SystemActions {
 
     private static final Thread NO_OP_SHUTDOWN_HOOK = new Thread(() -> {}, "Shutdown hook for testing");
 
-    @EntitlementTest(expectedAccess = ALWAYS_DENIED)
-    static void runtimeAddShutdownHook() {
-        Runtime.getRuntime().addShutdownHook(NO_OP_SHUTDOWN_HOOK);
-    }
-
     @EntitlementTest(expectedAccess = ALWAYS_DENIED)
     static void runtimeRemoveShutdownHook() {
         Runtime.getRuntime().removeShutdownHook(NO_OP_SHUTDOWN_HOOK);
diff --git a/libs/entitlement/src/main/java/org/elasticsearch/entitlement/initialization/HardcodedEntitlements.java b/libs/entitlement/src/main/java/org/elasticsearch/entitlement/initialization/HardcodedEntitlements.java
index 53e70f04..c2f5d7ea 100644
--- a/libs/entitlement/src/main/java/org/elasticsearch/entitlement/initialization/HardcodedEntitlements.java
+++ b/libs/entitlement/src/main/java/org/elasticsearch/entitlement/initialization/HardcodedEntitlements.java
@@ -162,8 +162,6 @@ class HardcodedEntitlements {
             )
         );
 
-        // conditionally add FIPS entitlements if FIPS only functionality is enforced
-        if (Booleans.parseBoolean(System.getProperty("org.bouncycastle.fips.approved_only"), false)) {
             // if custom trust store is set, grant read access to its location, otherwise use the default JDK trust store
             String trustStore = System.getProperty("javax.net.ssl.trustStore");
             Path trustStorePath = trustStore != null
@@ -177,7 +175,8 @@ class HardcodedEntitlements {
                     List.of(
                         new FilesEntitlement(List.of(FilesEntitlement.FileData.ofPath(trustStorePath, READ))),
                         new ManageThreadsEntitlement(),
-                        new OutboundNetworkEntitlement()
+                        new OutboundNetworkEntitlement(),
+                        new LoadNativeLibrariesEntitlement()
                     )
                 ),
                 new Scope(
@@ -187,9 +186,17 @@ class HardcodedEntitlements {
                         new FilesEntitlement(List.of(FilesEntitlement.FileData.ofBaseDirPath(LIB, READ))),
                         new ManageThreadsEntitlement()
                     )
+                ),
+                new Scope(
+                    "bc.rng.jent",
+                    // read to lib dir is required for checksum validation
+                    List.of(
+                        new FilesEntitlement(List.of(FilesEntitlement.FileData.ofBaseDirPath(LIB, READ))),
+                        new ManageThreadsEntitlement(),
+                        new LoadNativeLibrariesEntitlement()
+                    )
                 )
             );
-        }
         return serverScopes;
     }
 
diff --git a/libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementChecker.java b/libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementChecker.java
index a752f9c4..b9cac212 100644
--- a/libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementChecker.java
+++ b/libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementChecker.java
@@ -285,11 +285,6 @@ public class ElasticsearchEntitlementChecker implements EntitlementChecker {
         policyManager.checkChangeJVMGlobalState(callerClass);
     }
 
-    @Override
-    public void check$java_lang_Runtime$addShutdownHook(Class<?> callerClass, Runtime runtime, Thread hook) {
-        policyManager.checkChangeJVMGlobalState(callerClass);
-    }
-
     @Override
     public void check$java_lang_Runtime$removeShutdownHook(Class<?> callerClass, Runtime runtime, Thread hook) {
         policyManager.checkChangeJVMGlobalState(callerClass);
-- 
2.39.5 (Apple Git-154)

